### PR TITLE
Add type annotations for  pytorch_lightning/callbacks

### PR DESCRIPTION
**Pull request by hemildesai**
## What does this PR do?

Fixes part of https://github.com/PyTorchLightning/pytorch-lightning/issues/5023

Each file is in a separate commit, so might be easy to review commit by commit.

A lot of the places where I had to add `# type: ignore` are variables that were set to `None` earlier during initialization, but the code expects it to be a particular type because it is set a little later by a hook or something. Another common place where I had to add `# type: ignore` were external dependencies. 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by hemildesai** [comment]
> Hello @hemildesai! Thanks for updating this PR.
> 
> * In the file [`pytorch_lightning/callbacks/model_checkpoint.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/8ee02f8eee840fb2107bc41f7da1c7bbd10357b8/pytorch_lightning/callbacks/model_checkpoint.py):
> 
> > [Line 571:13](https://github.com/PyTorchLightning/pytorch-lightning/blob/8ee02f8eee840fb2107bc41f7da1c7bbd10357b8/pytorch_lightning/callbacks/model_checkpoint.py#L571): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> > [Line 572:13](https://github.com/PyTorchLightning/pytorch-lightning/blob/8ee02f8eee840fb2107bc41f7da1c7bbd10357b8/pytorch_lightning/callbacks/model_checkpoint.py#L572): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> > [Line 573:13](https://github.com/PyTorchLightning/pytorch-lightning/blob/8ee02f8eee840fb2107bc41f7da1c7bbd10357b8/pytorch_lightning/callbacks/model_checkpoint.py#L573): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> 
> ##### Comment last updated at 2020-12-13 08:10:31 UTC

W503 is in the ignore list in `.pep8speaks.yml` so not sure why it's showing up here.

File: pytorch_lightning/callbacks/base.py [hunk 1]
```diff
@@ -16,8 +16,16 @@
...
-
 import abc
+from typing import TYPE_CHECKING, Any, Dict, List, Union
+
+from torch.optim.optimizer import Optimizer
+
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.core.optimizer import LightningOptimizer
+
+if TYPE_CHECKING:
```
Comment (hemildesai) [inline pytorch_lightning/callbacks/base.py:27]: For context on this, see https://github.com/PyTorchLightning/pytorch-lightning/issues/5023#issuecomment-743745278

**Comment by tchaton** [comment]
Hey @hemildesai, great PR. Could you resolve the doc failing. Trainer doesn't exist :)

File: pytorch_lightning/trainer/attributes.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABC
+from typing import Any, Dict, List, Optional, Union
+
+
+class TrainerAttributes(ABC):
```
Comment (tchaton) [inline pytorch_lightning/trainer/attributes.py:18]: We are moving away from Mixin as hard to track and debug.

**Review by tchaton — COMMENTED** [review]
Great job ! Small changes required.

File: pytorch_lightning/trainer/attributes.py (hunk 2 above)
Comment (hemildesai) [reply pytorch_lightning/trainer/attributes.py:18]: What do you suggest for declaring attributes then? Directly on the `Trainer` class?

**Comment by hemildesai** [comment]
> Hello @hemildesai! Thanks for updating this PR.
> 
> * In the file [`pytorch_lightning/callbacks/model_checkpoint.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/8ee02f8eee840fb2107bc41f7da1c7bbd10357b8/pytorch_lightning/callbacks/model_checkpoint.py):
> 
> > [Line 571:13](https://github.com/PyTorchLightning/pytorch-lightning/blob/8ee02f8eee840fb2107bc41f7da1c7bbd10357b8/pytorch_lightning/callbacks/model_checkpoint.py#L571): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> > [Line 572:13](https://github.com/PyTorchLightning/pytorch-lightning/blob/8ee02f8eee840fb2107bc41f7da1c7bbd10357b8/pytorch_lightning/callbacks/model_checkpoint.py#L572): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> > [Line 573:13](https://github.com/PyTorchLightning/pytorch-lightning/blob/8ee02f8eee840fb2107bc41f7da1c7bbd10357b8/pytorch_lightning/callbacks/model_checkpoint.py#L573): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> 
> ##### Comment last updated at 2020-12-13 08:10:31 UTC

@tchaton `W503` is in the ignore list in `.pep8speaks.yml` so not sure why it's showing up here. I am using `Black` as the formatter and that seems to format it in this way.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -545,45 +553,47 @@ def _save_last_checkpoint(self, trainer, pl_module, ckpt_name_metrics, filepath)
...
             self._save_model(last_filepath, trainer, pl_module)
         if (
-                self.last_model_path
-                and self.last_model_path != last_filepath
-                and (self.save_top_k != -1 or self.save_last)
-                and trainer.is_global_zero
+            self.last_model_path
+            and self.last_model_path != last_filepath
+            and (self.save_top_k != -1 or self.save_last)
+            and trainer.is_global_zero
```
Comment (hemildesai) [inline pytorch_lightning/callbacks/model_checkpoint.py:573]: > Hello @hemildesai! Thanks for updating this PR.
> 
> * In the file [`pytorch_lightning/callbacks/model_checkpoint.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/8ee02f8eee840fb2107bc41f7da1c7bbd10357b8/pytorch_lightning/callbacks/model_checkpoint.py):
> 
> > [Line 571:13](https://github.com/PyTorchLightning/pytorch-lightning/blob/8ee02f8eee840fb2107bc41f7da1c7bbd10357b8/pytorch_lightning/callbacks/model_checkpoint.py#L571): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> > [Line 572:13](https://github.com/PyTorchLightning/pytorch-lightning/blob/8ee02f8eee840fb2107bc41f7da1c7bbd10357b8/pytorch_lightning/callbacks/model_checkpoint.py#L572): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> > [Line 573:13](https://github.com/PyTorchLightning/pytorch-lightning/blob/8ee02f8eee840fb2107bc41f7da1c7bbd10357b8/pytorch_lightning/callbacks/model_checkpoint.py#L573): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> 
> ##### Comment last updated at 2020-12-13 08:10:31 UTC

@tchaton `W503` is in the ignore list in `.pep8speaks.yml` so not sure why it's showing up here. I am using `Black` as the formatter and that seems to format it in this way.

**Comment by hemildesai** [comment]
@tchaton There's an issue with `sphinx-autodoc-typehints` not resolving forward references with `typing.TYPE_CHECKING`. The way they suggest to solve this is at https://github.com/agronholm/sphinx-autodoc-typehints#dealing-with-circular-imports

If solved using the suggested approach, `vscode` fails to resolve the type hints.

**Comment by Borda** [comment]
@hemildesai mind resolve conflict so we can let it land?

**Comment by hemildesai** [comment]
@Borda Can you look at my comment above and suggest the way forward? 

I'm happy to change the type references to 'pl.Trainer', but that won't resolve in some IDEs including VSCode

**Comment by Borda** [comment]
@hemildesai as it is a feature I am changing the target branch to feat 1.2, mind rebase pr and resolve conflicts? Thx your contribution and feel free to ping us back if you need something or need some help... 🐰
